spec/spec_helper.rb

```diff
@@ -1,80 +1,109 @@
+      "HostConfig" => {
+        "Privileged" => true,
+        "Binds" => ["#{`pwd`.chomp}:/workspace"],
+        "PublishAllPorts" => true,
+      },
+      "Env" => %w[
+        DRAUPNIR_DATABASE_URL=postgres://draupnir:draupnir@127.0.0.1/draupnir?sslmode=disable
+        DRAUPNIR_DATA_PATH=/draupnir
+        DRAUPNIR_ENVIRONMENT=test
+        DRAUPNIR_OAUTH_CLIENT_ID=the_client_id
+        DRAUPNIR_OAUTH_CLIENT_SECRET=the_client_secret
+        DRAUPNIR_OAUTH_REDIRECT_URL=https://server.com/oauth_callback
+        DRAUPNIR_PASSWORD=draupnir
+        DRAUPNIR_PORT=8443
+        DRAUPNIR_SHARED_SECRET=thesharedsecret
+        DRAUPNIR_TLS_CERTIFICATE_PATH=/etc/ssl/certs/draupnir_cert.pem
+        DRAUPNIR_TLS_PRIVATE_KEY_PATH=/etc/ssl/certs/draupnir_key.pem
+        DRAUPNIR_TRUSTED_USER_EMAIL_DOMAIN=@gocardless.com
+      ],
     )
+
+    draupnir.start!
+    draupnir.exec([BOOTSTRAP], &STREAMER)
+
+    new(draupnir)
   end
+  # rubocop:enable Metrics/MethodLength, Metrics/LineLength
 
-  def get(path)
-    RestClient::Request.execute(
-      verify_ssl: false,
-      method: :get,
-      url: "#{SERVER_ADDR}#{path}",
-      headers: {
-        content_type: JSON_CONTENT_TYPE,
-        authorization: "Bearer #{ACCESS_TOKEN}",
-        draupnir_version: DRAUPNIR_VERSION,
-      },
-    )
+  def initialize(draupnir)
+    @draupnir = draupnir
+    @host = URI.parse(Docker.connection.url).host || "127.0.0.1"
+    @port = @draupnir.json["NetworkSettings"]["Ports"]["8443/tcp"][0]["HostPort"]
   end
 
-  def delete(path)
+  delegate :remove, :exec, :store_file, to: :@draupnir
+
+  def request(method, path, payload = nil, headers = {})
     RestClient::Request.execute(
       verify_ssl: false,
-      method: :delete,
-      url: "#{SERVER_ADDR}#{path}",
+      method: method,
+      url: "https://#{@host}:#{@port}#{path}",
+      payload: payload&.to_json,
       headers: {
-        content_type: JSON_CONTENT_TYPE,
-        authorization: "Bearer #{ACCESS_TOKEN}",
-        draupnir_version: DRAUPNIR_VERSION,
-      },
+        content_type: "application/json",
+        authorization: "Bearer thesharedsecret",
+        draupnir_version: VERSION,
+      }.merge(headers),
     )
   end
 
-  def destroy_all_instances!
-    instances = JSON.parse(get("/instances"))["data"]
+  def alive?
+    JSON.parse(request(:get, "/health_check"))["status"] == "ok"
+  end
+
+  def destroy_all_instances
+    instances = JSON.parse(request(:get, "/instances"))["data"]
 
     instances.each do |instance|
-      delete("/instances/#{instance['id']}")
+      request(:delete, "/instances/#{instance['id']}")
     end
   end
 
-  def destroy_all_images!
-    images = JSON.parse(get("/images"))["data"]
+  def destroy_all_images
+    images = JSON.parse(request(:get, "/images"))["data"]
 
     images.each do |image|
-      delete("/images/#{image['id']}")
+      request(:delete, "/images/#{image['id']}")
     end
   end
+end
+
```

Comment:
Not a huge fan of having to refer to `@draupnir` everywhere in specs. Can we not just define the three http methods here?
```ruby
def client
  @client ||= Draupnir.create_from_container.tap { |c| raise "draupnir did not boot" unless c.alive? }
end

def get(path)
  client.request(:get, path)
end

def post(path, payload, headers={})
  client.request(:post, path, payload, headers)
end

def delete(path)
  client.request(:delete, path)
end
```

Reply:
How come you don't like the `@draupnir` in the specs? Can I argue devils advocate that it makes it very clear from just a glance which parts of the spec are interacting with our test instance?

I'm not attached either way but I slightly favour having one handle to a class, rather than many distinct methods.

Reply:
It's noisy and unhelpful - the entire test suite involves hitting the service with http calls, so it doesn't distinguish anything. `draupnir` is also not a helpful name, since it refers to the project as a whole.

Reply:
This also begs the question about how we handle booting and destroying the instance in tests- whether we want to pay the price to boot an instance each time or whether it's best to boot one and deal with it if we break that instance in the middle of a test suite.

Reply:
That's fair, I think it makes it quite difficult to see what you're actually doing when calling the `get/post/delete` methods though! Will make the change.

Reply:
On a side note, I think `@draupnir` is more honest in how it exposes the underlying mechanism to the test code, in that it makes it clear there is a global handle with state being passed around.

Reply:
It's a piece of state that doesn't change throughout the lifetime of the test suite though. This is exactly equivalent to spinning up some service (e.g. elasticsearch) before running your integration tests.